cumulus/parachains/runtimes/bridge-hubs/common/tests/tests.rs

```diff
@@ -0,0 +1,140 @@
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg(test)]
+use bridge_hub_common::DenyExportMessageFrom;
+use frame_support::{
+	assert_err, assert_ok, parameter_types,
+	traits::{Equals, Everything, EverythingBut, ProcessMessageError},
+};
+use xcm::prelude::{
+	AliasOrigin, All, AssetFilter, DepositReserveAsset, Ethereum, ExportMessage, Here, Instruction,
+	Location, NetworkId, NetworkId::Polkadot, Parachain, Weight, Wild,
+};
+use xcm_builder::{DenyReserveTransferToRelayChain, DenyThenTry, TakeWeightCredit};
+use xcm_executor::traits::{Properties, ShouldExecute};
+
+parameter_types! {
+	pub AssetHubLocation: Location = Location::new(1, Parachain(1000));
+	pub ParachainLocation: Location = Location::new(1, Parachain(2000));
+	pub EthereumNetwork: NetworkId = Ethereum { chain_id: 1 };
+}
+
+#[test]
+fn deny_export_message_from_source_other_than_asset_hub_should_work() {
+	pub type Barrier = DenyThenTry<
+		(
+			DenyReserveTransferToRelayChain,
+			DenyExportMessageFrom<EverythingBut<Equals<AssetHubLocation>>, Equals<EthereumNetwork>>,
+		),
+		TakeWeightCredit,
+	>;
+
+	let mut xcm: Vec<Instruction<()>> = vec![
+		AliasOrigin(Here.into()),
+		ExportMessage {
+			network: EthereumNetwork::get(),
+			destination: Here,
+			xcm: Default::default(),
+		},
+	];
+
+	let result = Barrier::should_execute(
+		&ParachainLocation::get(),
+		&mut xcm,
+		Weight::zero(),
+		&mut Properties { weight_credit: Weight::zero(), message_id: None },
+	);
+
+	assert_err!(result, ProcessMessageError::Unsupported);
+}
+
+#[test]
+fn allow_export_message_from_asset_hub_should_work() {
+	pub type Barrier = DenyThenTry<
+		(
+			DenyReserveTransferToRelayChain,
+			DenyExportMessageFrom<EverythingBut<Equals<AssetHubLocation>>, Equals<EthereumNetwork>>,
+		),
+		TakeWeightCredit,
+	>;
+
+	let mut xcm: Vec<Instruction<()>> = vec![
+		AliasOrigin(Here.into()),
+		ExportMessage {
+			network: EthereumNetwork::get(),
+			destination: Here,
+			xcm: Default::default(),
+		},
+	];
+
+	let result = Barrier::should_execute(
+		&AssetHubLocation::get(),
+		&mut xcm,
+		Weight::zero(),
+		&mut Properties { weight_credit: Weight::zero(), message_id: None },
+	);
+
+	assert_ok!(result);
+}
+
+#[test]
+fn allow_export_message_from_source_other_than_asset_hub_if_destination_other_than_ethereum() {
+	pub type Barrier = DenyThenTry<
+		(
+			DenyReserveTransferToRelayChain,
+			DenyExportMessageFrom<EverythingBut<Equals<AssetHubLocation>>, Equals<EthereumNetwork>>,
+		),
+		TakeWeightCredit,
+	>;
+
+	let mut xcm: Vec<Instruction<()>> = vec![
+		AliasOrigin(Here.into()),
+		ExportMessage { network: Polkadot, destination: Here, xcm: Default::default() },
+	];
+
+	let result = Barrier::should_execute(
+		&ParachainLocation::get(),
+		&mut xcm,
+		Weight::zero(),
+		&mut Properties { weight_credit: Weight::zero(), message_id: None },
+	);
+
+	assert_ok!(result);
+}
+
+#[test]
+fn deny_reserver_transfer_to_relaychain_does_not_break() {
```

Comment:
please, remove this test from here and again move it to the https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-builder/src/tests/barriers.rs and keep it as a dedicated `DenyReserveTransferToRelayChain` without chaining with other barriers

Reply:
https://github.com/paritytech/polkadot-sdk/pull/7169/commits/58d79909011fb0f6173542ff779d210835338e49